KOGITO-8740: data-index documentation (#409)
kiegroup/kogito-docs: serverlessworkflow/modules/ROOT/pages/data-index/data-index-quarkus-extension.adoc

In the same way as the {data_index_ref} service, there is a specific addon for each type of persistence you want to work with. Currently, you can find {data_index_ref} addons for: link:{postgresql_url}[PostgreSQL], link:{infinispan_url}[Infinispan], and link:{mongo_url}[MongoDB]

[IMPORTANT]
====
The {data_index_ref} addon distribution added to the workflow must match the workflow's persistence type regarding the indexed data, which can be stored in the same database.
The addon will share the data source used by the workflow where it is added, and it will create separate tables for that purpose.
====

When any of the {data_index_ref} addons is added:

* The communication with the workflow is direct, the workflow data is *not* transmitted or consumed through events, they are stored directly in the configured database.
There is no need to configure the events connection for this purpose.
* A new GraphQL endpoint is added to perform queries and management operations

[[data-index-ext-use]]
== Adding {data_index_ref} extension to a workflow application

> You can add the {data_index_ref} quarkus extension as an Addon:

Reviewer: ```suggestion
You can add the {data_index_ref} quarkus extension as an addon.
```

// External pages
:kogito_sw_timeouts_showcase_embedded_example_url: {kogito_sw_examples_url}/serverless-workflow-timeouts-showcase-embedded
:kogito_sw_timeouts_showcase_embedded_example_application_properties_url: {kogito_sw_timeouts_showcase_embedded_example_url}/src/main/resources/application.properties
:infinispan_url: https://infinispan.org/
:mongo_url: https://www.mongodb.com/
:postgresql_url: https://www.postgresql.org/

//Common constants
:data_index_ref: Data Index
:workflow_instance: process instance
:workflow_instances: {workflow_instance}s

This document describes how you add the {data_index_ref} features to your runtime service. You simply need to add the {data_index_ref} extension to the runtime and
the workflow related data will be collected and stored in the database, enabling the GraphQL endpoint to execute queries and perform management operations over the {workflow_instances}.
The example described in this document is based on the link:{kogito_sw_timeouts_showcase_embedded_example_url}[`serverless-workflow-timeouts_showcase_embedded`] example application.

[[data-index-ext-overview]]
== Overview

> The {data_index_ref} service has been designed to store and manage data from different workflows that can be hosted in different runtimes.

Author: I mean the data-index was thought to index data from different workflows that can be placed in different workflow applications. I understand that one application can deploy many workflows but the data-index is expecting that the event with the data, incorporates the endpoint related with  the workflow instance, allowing to connect the instances independently on where they are deployed.

> The {data_index_ref} service has been designed to store and manage data from different workflows that can be hosted in different runtimes.

Reviewer 2: So what about this rewording?

```suggestion
The {data_index_ref} service has been designed to store and manage data from different workflow instances.
```

I believe we can refrain from saying "hosted in different workflow/applications/runtimes".

> = Data Index Quarkus Extension

Reviewer 3: ```suggestion
= Data Index Quarkus extension
```

:compat-mode!:
// Metadata:
:description: Data Index Service to allow to index and query audit data in {product_name}
:keywords: kogito, workflow, serverless, data, dataindex, data-index, index, service, extension, quarkus
// Referenced documentation pages
:persistence_with_postgresql_guide: xref:persistence/persistence-with-postgresql.adoc
:data-index-core-concepts_guide: xref:data-index/data-index-core-concepts.adoc
:getting_started_create_first_workflow_guide: xref:getting-started/create-your-first-workflow-service.adoc
// External pages
:kogito_sw_timeouts_showcase_embedded_example_url: {kogito_sw_examples_url}/serverless-workflow-timeouts-showcase-embedded
:kogito_sw_timeouts_showcase_embedded_example_application_properties_url: {kogito_sw_timeouts_showcase_embedded_example_url}/src/main/resources/application.properties
:infinispan_url: https://infinispan.org/
:mongo_url: https://www.mongodb.com/
:postgresql_url: https://www.postgresql.org/

This document describes how you add the Data Index features to your runtime service. You simply need to add the Data Index extension to the runtime and

> all the audit data will be collected and stored in the database, and the graphQL endpoint will be available to execute queries and perform management operation over the workflows instances.

Reviewer 4: ```suggestion
all the audit data will be collected and stored in the database, and the GraphQL endpoint will be available to execute queries and perform management operations over the workflows instances.
```

> all the audit data will be collected and stored in the database, and the graphQL endpoint will be available to execute queries and perform management operation over the workflows instances.

Reviewer 3: ```suggestion
all the audit data will be collected and stored in the database, and the graphQL endpoint will be available to execute queries and perform management operations over the workflows instances.
```

The example described in this document is based on the link:{kogito_sw_timeouts_showcase_embedded_example_url}[`serverless-workflow-timeouts_showcase_embedded`] example application.

[[data-index-ext-overview]]
== Overview

> The Data Index service has been designed to store and manage audit data from different workflows that may be hosted in different runtimes.

Reviewer 4: ```suggestion
The Data Index service has been designed to store and manage audit data from different workflows that different runtimes may execute.
```

> The Data Index service has been designed to store and manage audit data from different workflows that may be hosted in different runtimes.

Reviewer 3: ```suggestion
The Data Index service is designed to store and manage audit data from different workflows that can be hosted in different runtimes.
```

> Communication with the service is through events that contain the audit data and the service is responsible for storing them and exposing a GRAPHQL endpoint to allow queries and maintenance operations on the different instances of Workflows.

Reviewer 4: ```suggestion
Communication with the service is through events that contain the audit data, and the service is responsible for storing them and exposing a GraphQL endpoint to allow queries and maintenance operations on the different workflow instances.
```

> Sometimes, to avoid deploying the service separately, it could be interesting to have the audit storage functionality and the query possibility embedded in the same application.

Reviewer 4: ```suggestion
In certain use cases, to avoid deploying the service separately, it could be beneficial to have the audit storage functionality and the query capabilities, embedded in the same application.
```

> Sometimes, to avoid deploying the service separately, it could be interesting to have the audit storage functionality and the query possibility embedded in the same application.

Reviewer 3: I have another suggestion as well :) 

```suggestion
In specific use cases, to avoid deploying the service separately, it could be useful to have the audit storage functionality and the query capabilities embedded in the same application.
```

> For this purpose, we have the Quarkus Data Index extension that can be plugged to any workflow application and incorporates the Data Index functionality into the same application with no need of an external Data Index Service.

Reviewer 4: ```suggestion
For this purpose, the Quarkus Data Index extension can be added to any workflow application and incorporates the Data Index functionality into the same application without needing an external Data Index Service.
```

> These extensions can be found distributed as Kogito addons: `kogito-addons-quarkus-data-index`.

Reviewer 4: mention kogito-addons-quarkus-data-index-postgresql or others?

[NOTE]
====
The Data Index extensions are provided as addons for each kind of supported persistence relying on the link:{quarkus_guides_base_url}/writing-extensions[Quarkus extensions] mechanism.
====

Once one of these `kogito-addons-quarkus-data-index` addons is added to a Workflow application, that service incorporates the functionality to store all the audit data and also incorporates the GraphQL endpoint to perform queries and management operations.

In the same way as the Data Index Service, there is a specific addon for each type of persistence you want to work with. Currently, we can find Data Index addons for: link:{postgresql_url}[PostgreSQL]
link:{infinispan_url}[Infinispan] and  link:{mongo_url}[MongoDB]

[IMPORTANT]
====

> The `kogito-addons-quarkus-data-index` addon distribution added to the application should match with the application kind of persistence, regarding the audit data will be stored in the same database.

Reviewer 3: ```suggestion
The `kogito-addons-quarkus-data-index` addon distribution added to the application must match the application's kind of persistence, regarding the audit data, which can be stored in the same database.
```

> The addon will share the data-source used by the application where it is added, and it will create separated audit tables for that purpose.

Reviewer 3: ```suggestion
The addon will share the data source used by the application where it is added, and it will create separate audit tables for that purpose.
```

====

> When any of the data index addons is added:

Reviewer 4: ```suggestion
When any of the Data Index addons is added:
```

> * The communication with the workflow application is direct, the audit data are *not* transmitted or consumed through events, they are stored directly in the configured DB.

Reviewer 3: Here DB means database, right? 
```suggestion
* The communication with the workflow application is direct, the audit data are *not* transmitted or consumed through events, they are stored directly in the configured database.
```

Author: yes, replaced!

There is no need to configure the events connection for this purpose.
* A new GraphQL endpoint is added to perform queries and management operations

[[data-index-ext-use]]
== Adding Data Index extension to a workflow application

You can add the Data Index quarkus extension as an Addon:

.Prerequisites
* Your workflow application is working and has persistence enabled.
+
For more information about creating a workflow application, see {getting_started_create_first_workflow_guide}[Creating your first workflow service]. You also can find more details about enabling persistence in {persistence_with_postgresql_guide}[Running a workflow service using PostgreSQL]

.Procedure
. Add the required Data Index addon dependencies to the `pom.xml` file of your workflow application:
+
--
.Add Data Index Addon dependencies to `pom.xml` file
[source,xml]
----
<dependency>
    <groupId>org.kie.kogito</groupId>
    <artifactId>kogito-addons-quarkus-data-index-postgresql</artifactId>
</dependency>
----

[NOTE]
====

> In the example the Workflow Application is using PostgreSQL persistence, that's the reason to include the kogito-addons-quarkus-data-index-postgresql,

Reviewer 3: ```suggestion
In the example the Workflow Application is using PostgreSQL persistence, that's the reason for including the kogito-addons-quarkus-data-index-postgresql,
```

> but this addon is distributed ready to work with other types of persistence:

Reviewer 3: ```suggestion
but this addon is distributed and ready to work with other types of persistence:
```

* kogito-addons-quarkus-data-index-inmemory (inmemory PostgreSQL)
* kogito-addons-quarkus-data-index-postgresql
* kogito-addons-quarkus-data-index-infinispan
* kogito-addons-quarkus-data-index-mongodb
====

--

. Configure the Data Index Addon properties.
+
--

> Prepare the application to run in dev mode, avoiding to start the DataIndex dev service and specify if the GraphQL UI needs to be available.

Reviewer 3: ```suggestion
Prepare the application to run in dev mode, avoid starting the DataIndex dev service, and specify if the GraphQL UI needs to be available.
```
